lib/redis_mutex/lock.ex

```diff
@@ -1,118 +1,131 @@
+    @callback child_spec(opts :: Keyword.t()) :: Supervisor.child_spec()
 
-      block_value = unquote(clause)
+    @callback start_link(start_options :: RedisMutex.start_options()) ::
+                {:ok, pid()} | {:error, any()}
 
-      RedisMutex.Lock.unlock(key, uuid)
+    @callback with_lock(key :: String.t(), do: clause :: term()) :: any()
 
-      block_value
+    @callback with_lock(key :: String.t(), timeout :: integer(), do: clause :: term()) :: any()
+
+    @callback with_lock(key :: String.t(), timeout :: integer(), expiry :: integer(),
+                do: clause :: term()
+              ) ::
+                any()
+
+    @spec child_spec(Keyword.t()) :: Supervisor.child_spec()
+    def child_spec(opts) do
+      %{
+        id: __MODULE__,
+        start: {__MODULE__, :start_link, opts},
+        type: :worker,
+        restart: :permanent,
+        shutdown: 500
+      }
     end
-  end
 
-  @doc """
-  This function takes in a key, unique string, and a timeout in milliseconds.
-  It will call itself recursively until it is able to set a lock
-  or the timeout expires.
-  """
-  def take_lock(key, uuid, timeout \\ @default_timeout, expiry \\ @default_expiry, finish \\ nil)
+    @spec start_link(RedisMutex.start_options()) :: :ignore | {:error, any} | {:ok, pid}
+    def start_link(start_options \\ []) do
+      {redis_url, redix_opts} = Keyword.pop(start_options, :redis_url)
 
-  def take_lock(key, uuid, timeout, expiry, nil) do
-    finish = DateTime.add(DateTime.utc_now(), timeout, :millisecond)
-    take_lock(key, uuid, timeout, expiry, finish)
-  end
+      case redis_url do
+        redis_url when is_binary(redis_url) ->
+          Redix.start_link(redis_url, name: @name, sync_connect: true)
 
-  def take_lock(key, uuid, timeout, expiry, finish) do
-    if DateTime.compare(finish, DateTime.utc_now()) == :lt do
-      raise RedisMutex.Error, message: "Unable to obtain lock."
+        nil ->
+          [name: @name, sync_connect: true]
+          |> Keyword.merge(redix_opts)
+          |> Redix.start_link()
+      end
     end
 
-    if !lock(key, uuid, expiry) do
+    @spec with_lock(String.t(), integer(), integer(), do: term()) :: any()
+    defmacro with_lock(key, timeout \\ @default_timeout, expiry \\ @default_expiry, do: clause) do
+      quote do
+        key = unquote(key)
+        timeout = unquote(timeout)
+        expiry = unquote(expiry)
+        uuid = Uniq.UUID.uuid1()
+
+        RedisMutex.Lock.take_lock(key, uuid, timeout, expiry)
+
+        block_value = unquote(clause)
+
+        RedisMutex.Lock.unlock(key, uuid)
+
+        block_value
+      end
+    end
+
+    @doc """
+    This function takes in a key, which should be a unique string,
+    an optional timeout in milliseconds and an optional expiry,
+    also in milliseconds.
+    It will call itself recursively until it is able to set a lock
+    or the timeout expires.
+    """
+    def take_lock(
```

Comment:
Needs typespec

Can this be private?

Reply:
If I change the call inside the `with_lock` macro from `RedisMutex.Lock.take_lock` to `take_lock` then they can be private.